#520 Introduce JS context option 'print-no-newline'

e-hubert opened the pull request:
If this new option is configured the builtin print function will not
append a trailing newline. By default (without explicit configuration)
there will be no functional change.

e-hubert commented:
I just quickly created a PR without any test coverage to gather feedback on the approach. Of course I'm going to add at least a unit test for the new feature, if the approach does make sense.

iamstolis commented:
> I just quickly created a PR without any test coverage to gather feedback on the approach.

Your changes look good. I added just one trivial comment.

> Of course I'm going to add at least a unit test for the new feature, if the approach does make sense.

I am looking forward to the unit test. Thank you in advance!

wirthi commented:
I don't fully understand your use-case, but we also use `JSRuntime.LINE_SEPARATOR` in `JSConsoleCountNode`. Should we also make it configurable there? The other `Console` builtins (`Console.log` etc.) are deferring to the `JSGlobalPrintNode` so ARE using your changes.

iamstolis commented:
> I don't fully understand your use-case

It is mainly a Nashorn-compatibility feature (Nashorn has a similar option). So, it is crucial that `print` is affected. I don't have a strong opinion on whether `console.log` & co. should be affected but it seems like a good idea for the sake of consistency and simplicity of the implementation.

> but we also use JSRuntime.LINE_SEPARATOR in JSConsoleCountNode.

Good point. If we keep the current state where `console.log` & co. are affected then the newline in `console.count()` should be affected too.

e-hubert commented:
@iamstolis, @wirthi: Thanks for your input! I looked again at it and also compared to the behavior in the Nashorn engine and think it would be best to limit the change to the print function only and let the ConsoleBuiltins unaffected, because there it does not make sense to me for many use cases. 
I think, we can easily achieve this by introducing an addition constructor arg to JSGlobalPrintNode. What do you think? If this is OK, shall we go fo an overloaded constructor keeping the existing one for binary compatibility? If it is fine with you, I can push another take on this.

iamstolis commented:
> I looked again at it and also compared to the behavior in the Nashorn engine and think it would be best to limit the change to the print function only and let the ConsoleBuiltins unaffected

That's a valid approach. As I mentioned already, I don't have a strong opinion on whether `ConsoleBuiltins` should be affected or not. My only constraints are that either all or none of them should be affected by the new option and that both `print` and `printErr` should be affected.

> I think, we can easily achieve this by introducing an addition constructor arg to JSGlobalPrintNode. What do you think? If this is OK, shall we go for an overloaded constructor keeping the existing one for binary compatibility?

Feel free to add a new parameter to the existing constructor. There is no need to keep the binary compatibility. `JSGlobalPrintNode` is not a public API. It is an internal implementation detail of `graal-js`.

e-hubert commented:
Alright @iamstolis, thanks for confirming. I just did not want to take such a change too light-headed. Pushed the change as described and also added the corresponding tests to printErr to manifest the intentional change. We would of course have the chance to keep printErr unaffected - but you strongly preferred those to be consistent, so I implemented it this way.

iamstolis commented:
> Pushed the change as described

Your changes look good to me.

> and also added the corresponding tests to printErr to manifest the intentional change.

Thanks for the tests.
